### PR TITLE
[BE] Fix MPS build warnings

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83049
* __->__ #83048

Mostly get rid of unused variables, but also:
- Hide `-[MPSGraphTensorData printNDArray]` behind undefined method
  access
- Rename several methods in ScatterGather/TriangularOps:
   - `scatterAlongAxisWithDataTensor:`->`scatterAlongAxis:withDataTensor:`
   - `gatherAlongAxisWithUpdatesTensor:`->`gatherAlongAxis:withUpdatesTensor:`
   - `getCoordinateValueWithShapeTensor:`->`coordinateAlongAxisTensor:withShapeTensor:` 
- Add `-Wno-unguarded-availability-new` to suppress 12.3+ availability warnings

### CODE REVIEW

File: aten/src/ATen/native/mps/MPSGraph+CoolMethods.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+// Nor in MacOSX12.3.sdk headers
+// TODO: Do not call to undocumented methods
+
+@interface MPSGraph(CoolMethods)
+-(MPSGraphTensor *) gatherAlongAxisWithUpdatesTensor:(MPSGraphTensor *) updateTensor
+                                       indicesTensor:(MPSGraphTensor *) indicesTensor
+                                                axis:(NSInteger)dim
+                                                name:(NSString * _Nullable)name;
+
+-(MPSGraphTensor *) getCoordinateValueWithShapeTensor:(MPSGraphTensor *) shapeTensor
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/MPSGraph+CoolMethods.h:18]: @malfet, can you please replace these methods with the Public SDK ones, also please remove this file. We won't need CoolMethods 😄 .
```
-(MPSGraphTensor *) coordinateAlongAxisTensor:(MPSGraphTensor *) axisTensor
                              withShapeTensor:(MPSGraphTensor *) shapeTensor
                                         name:(NSString * _Nullable) name
```
Note the axisTensor and shapeTensor are swapped.

File: aten/src/ATen/native/mps/MPSGraph+CoolMethods.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+NS_ASSUME_NONNULL_BEGIN
+
+// ScatterGather.mm and TriangularOps.mm call to those
+// But they can not be found in https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraph?language=objc
+// Nor in MacOSX12.3.sdk headers
+// TODO: Do not call to undocumented methods
+
+@interface MPSGraph(CoolMethods)
+-(MPSGraphTensor *) gatherAlongAxisWithUpdatesTensor:(MPSGraphTensor *) updateTensor
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/MPSGraph+CoolMethods.h:13]: For this the the right API is:

```
-(MPSGraphTensor *) gatherAlongAxis:(NSInteger) axis
                  withUpdatesTensor:(MPSGraphTensor *) updatesTensor
                      indicesTensor:(MPSGraphTensor *) indicesTensor
                               name:(NSString * _Nullable) name
```

File: aten/src/ATen/native/mps/MPSGraph+CoolMethods.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+                                       indicesTensor:(MPSGraphTensor *) indicesTensor
+                                                axis:(NSInteger)dim
+                                                name:(NSString * _Nullable)name;
+
+-(MPSGraphTensor *) getCoordinateValueWithShapeTensor:(MPSGraphTensor *) shapeTensor
+                                           axisTensor:(MPSGraphTensor *) axisTensor
+                                                 name:(NSString * _Nullable) name;
+
+-(MPSGraphTensor *) scatterAlongAxisWithDataTensor:(MPSGraphTensor *) dataTensor
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/MPSGraph+CoolMethods.h:22]: For this the right API is:
```
-(MPSGraphTensor *) scatterAlongAxis:(NSInteger) axis
                      withDataTensor:(MPSGraphTensor *) dataTensor
                       updatesTensor:(MPSGraphTensor *) updatesTensor
                       indicesTensor:(MPSGraphTensor *) indicesTensor
                                mode:(MPSGraphScatterMode) mode
                                name:(NSString * _Nullable) name
```

**Review by kulinseth — CHANGES_REQUESTED** [review]

**Comment by malfet** [comment]
@kulinseth updated selector names per your suggestions

File: aten/src/ATen/mps/MPSGuardImpl.mm [hunk 4]
```diff
@@ -9,9 +9,6 @@
   void MPSGuardImpl::createEvent(
     mpsEvent_t* event,
     const EventFlag flag) const {
-    id<MTLDevice> mtl_device = MPSDevice::getInstance()->device();
```
Comment (albanD) [inline aten/src/ATen/mps/MPSGuardImpl.mm:12]: Wait what is this function doing now? Both args are ignored and it is a no-op??

**Review by albanD — APPROVED** [review]
Very nice cleanup.
Thanks for taking the time to do this Nikita!

File: aten/src/ATen/mps/MPSGuardImpl.mm (hunk 4 above)
Comment (malfet) [reply aten/src/ATen/mps/MPSGuardImpl.mm:12]: I don't know what its doing, but even before the change it was a no-op, wasn't it?
But I guess it was supposed to do something like `*event = new mpsEvent(flag);`?

Comment (albanD) [reply aten/src/ATen/mps/MPSGuardImpl.mm:12]: It was indeed a noop before so this is ok. But we should fix this for sure!

**Pull request closed without merging**